### PR TITLE
bevy_asset: Improve `NestedLoader` API

### PR DESCRIPTION
**Pull request by aecsocket**
# Objective

The `NestedLoader` API as it stands right now is somewhat lacking:

- It consists of several types `NestedLoader`, `UntypedNestedLoader`, `DirectNestedLoader`, and `UntypedDirectNestedLoader`, where a typestate pattern on `NestedLoader` would be make it more obvious what it does, and allow centralising the documentation
- The term "untyped" in the asset loader code is overloaded. It can mean either:
  - we have literally no idea what the type of this asset will be when we load it (I dub this "unknown type")
  - we know what type of asset it will be, but we don't know it statically - we only have a TypeId (I dub this "dynamic type" / "erased")
- There is no way to get an `UntypedHandle` (erased) given a `TypeId`

## Solution

Changes `NestedLoader` into a type-state pattern, adding two type params:
- `T` determines the typing
  - `StaticTyped`, the default, where you pass in `A` statically into `fn load<A>() -> ..`
  - `DynamicTyped`, where you give a `TypeId`, giving you a `UntypedHandle`
  - `UnknownTyped`, where you have literally no idea what type of asset you're loading, giving you a `Handle<LoadedUntypedAsset>`
- `M` determines the "mode" (bikeshedding TBD, I couldn't come up with a better name)
  - `Deferred`, the default, won't load the asset when you call `load`, but it does give you a `Handle` to it (this is nice since it can be a sync fn)
  - `Immediate` will load the asset as soon as you call it, and give you access to it, but you must be in an async context to call it

Changes some naming of internals in `AssetServer` to fit the new definitions of "dynamic type" and "unknown type". Note that I didn't do a full pass over this code to keep the diff small. That can probably be done in a new PR - I think the definiton I laid out of  unknown type vs. erased makes it pretty clear where each one applies.

<details>
<summary>Old issue</summary>

The only real problem I have with this PR is the requirement to pass in `type_name` (from `core::any::type_name`) into Erased. Users might not have that type name, only the ID, and it just seems sort of weird to *have* to give an asset type name. However, the reason we need it is because of this:
```rs
    pub(crate) fn get_or_create_path_handle_erased(
        &mut self,
        path: AssetPath<'static>,
        type_id: TypeId,
        type_name: &str,
        loading_mode: HandleLoadingMode,
        meta_transform: Option<MetaTransform>,
    ) -> (UntypedHandle, bool) {
        let result = self.get_or_create_path_handle_internal(
            path,
            Some(type_id),
            loading_mode,
            meta_transform,
        );
        // it is ok to unwrap because TypeId was specified above
        unwrap_with_context(result, type_name).unwrap()
    }

pub(crate) fn unwrap_with_context<T>(
    result: Result<T, GetOrCreateHandleInternalError>,
    type_name: &str,
) -> Option<T> {
    match result {
        Ok(value) => Some(value),
        Err(GetOrCreateHandleInternalError::HandleMissingButTypeIdNotSpecified) => None,
        Err(GetOrCreateHandleInternalError::MissingHandleProviderError(_)) => {
            panic!("Cannot allocate an Asset Handle of type '{type_name}' because the asset type has not been initialized. \
                    Make sure you have called app.init_asset::<{type_name}>()")
        }
    }
}
```
This `unwrap_with_context` is literally the only reason we need the `type_name`. Potentially, this can be turned into an `impl Into<Option<&str>>`, and output a different error message if the type name is missing. Since if we are loading an asset where we only know the type ID, by definition we can't output that error message, since we don't have the type name. I'm open to suggestions on this.

</details>

## Testing

Not sure how to test this, since I kept most of the actual NestedLoader logic the same. The only new API is loading an `UntypedHandle` when in the `DynamicTyped, Immediate` state.

## Migration Guide

Code which uses `bevy_asset`'s `LoadContext::loader` / `NestedLoader` will see some naming changes:
- `untyped` is replaced by `with_unknown_type`
- `with_asset_type` is replaced by `with_static_type`
- `with_asset_type_id` is replaced by `with_dynamic_type`
- `direct` is replaced by `immediate` (the opposite of "immediate" is "deferred")

### CODE REVIEW

File: crates/bevy_asset/src/server/mod.rs [hunk 1]
```diff
@@ -401,6 +401,44 @@ impl AssetServer {
         handle
     }
 
+    pub(crate) fn load_erased_with_meta_transform<'a, G: Send + Sync + 'static>(
```
Comment (aecsocket) [inline crates/bevy_asset/src/server/mod.rs:404]: This fn is effectively a carbon copy of `load_with_meta_transform` but using `get_or_create_path_handle_erased`. I don't like the code duplication - should this be combined into one fn or left separate?

File: crates/bevy_asset/src/loader_builders.rs [hunk 2]
```diff
@@ -25,28 +25,164 @@ impl ReaderRef<'_> {
...
+mod sealed {
+    pub trait Typing {}
+
+    pub trait Mode {}
+}
+
+/// [`NestedLoader`] will be provided the type of asset as a type parameter on
+/// [`load`].
+///
+/// [`load`]: Self::load
```
Comment (andriyDev) [inline crates/bevy_asset/src/loader_builders.rs:134]: 1) Does rustdoc "redirect" these links for you? Just for my own learning (and can you verify this)
2) I think `Self::load` is undefined here.

File: crates/bevy_asset/src/loader_builders.rs [hunk 3]
```diff
@@ -25,28 +25,164 @@ impl ReaderRef<'_> {
...
 
-impl<'ctx, 'builder> NestedLoader<'ctx, 'builder> {
-    pub(crate) fn new(
-        load_context: &'builder mut LoadContext<'ctx>,
-    ) -> NestedLoader<'ctx, 'builder> {
+impl sealed::Typing for Typed {}
+
+/// [`NestedLoader`] has been configured with info on what type of asset to load
+/// at runtime.
+pub struct Erased<'builder> {
```
Comment (andriyDev) [inline crates/bevy_asset/src/loader_builders.rs:143]: As I stated in the Discord, I would prefer to just call this DynamicType, since erased is a pretty technical term IMO.

But ideally I'd like someone else's opinion since mine is weakly held.

File: crates/bevy_asset/src/loader_builders.rs [hunk 4]
```diff
@@ -74,47 +210,108 @@ impl<'ctx, 'builder> NestedLoader<'ctx, 'builder> {
...
+    /// When [`load`]ing, we will infer what type of asset to load from
+    /// metadata.
+    ///
+    /// [`load`]: Self::load
     #[must_use]
-    pub fn direct<'c>(self) -> DirectNestedLoader<'ctx, 'builder, 'c> {
-        DirectNestedLoader {
-            base: self,
-            reader: None,
+    pub fn of_unknown_type(self) -> NestedLoader<'ctx, 'builder, UnknownType, M> {
```
Comment (andriyDev) [inline crates/bevy_asset/src/loader_builders.rs:260]: I think I'd prefer `unknown_type` rather than `of_unknown_type`? Makes it match `erased` and `typed` more closely, idk.

File: crates/bevy_asset/src/loader_builders.rs [hunk 5]
```diff
@@ -74,47 +210,108 @@ impl<'ctx, 'builder> NestedLoader<'ctx, 'builder> {
...
+    /// a dependency of this asset.
+    ///
+    /// This requires you to know the type of asset statically.
+    /// - If you have runtime info for what type of asset you're loading (e.g. a
+    ///   [`TypeId`]), use [`erased`].
+    /// - If you do not know at all what type of asset you're loading, use
+    ///   [`unknown_type`].
+    ///
+    /// [`erased`]: Self::erased
+    /// [`unknown_type`]: Self::unknown_type
```
Comment (andriyDev) [inline crates/bevy_asset/src/loader_builders.rs:314]: This line doesn't match, so if you don't rename `of_unknown_type`, fixup this doc.

File: crates/bevy_asset/src/server/mod.rs (hunk 1 above)
Comment (andriyDev) [reply crates/bevy_asset/src/server/mod.rs:404]: I think just deduping the `should_load` part should be enough. The rest of it is pretty small.

File: crates/bevy_asset/src/loader_builders.rs (hunk 2 above)
Comment (aecsocket) [reply crates/bevy_asset/src/loader_builders.rs:134]: 1. If you put
```
[foo]: bar
```
in its own paragraph in markdown, it'll redirect `[foo]` to `bar`.
image

2. You're right, this is a copy paste error from me moving docs around. (Also, this [gets caught by CI](https://github.com/bevyengine/bevy/actions/runs/11087621869/job/30806525049))

File: crates/bevy_asset/src/loader_builders.rs (hunk 4 above)
Comment (aecsocket) [reply crates/bevy_asset/src/loader_builders.rs:260]: Makes sense to me, I'll rename to `unknown_type`

File: crates/bevy_asset/src/loader_builders.rs (hunk 5 above)
Comment (aecsocket) [reply crates/bevy_asset/src/loader_builders.rs:314]: Left it as `unknown_type`

File: crates/bevy_asset/src/loader_builders.rs (hunk 3 above)
Comment (aecsocket) [reply crates/bevy_asset/src/loader_builders.rs:143]: After some thought, I agree with `DynamicType` more. It's a bit less concise than `Erased`, but it does match the semantics of `dyn`.

I'll rename `Typed` to `StaticTyped`, `Erased` to `DynamicTyped`, and `UnknownType` to `UnknownTyped`.

File: crates/bevy_asset/src/loader_builders.rs (hunk 4 above)
Comment (aecsocket) [reply crates/bevy_asset/src/loader_builders.rs]: Renamed to `with_unknown_type` to keep it in line with:
- `with_settings`
- `with_[static|dynamic]_type`

File: crates/bevy_asset/src/server/mod.rs [hunk 6]
```diff
@@ -401,6 +395,51 @@ impl AssetServer {
...
+        );
+
+        if should_load {
+            let task = self.spawn_load_task(handle.clone(), path, guard);
+
+            #[cfg(not(any(target_arch = "wasm32", not(feature = "multi_threaded"))))]
+            infos.pending_tasks.insert(handle.id(), task);
+
+            #[cfg(any(target_arch = "wasm32", not(feature = "multi_threaded")))]
+            task.detach();
```
Comment (aecsocket) [inline crates/bevy_asset/src/server/mod.rs:422]: I couldn't deduplicate this block because it uses the `infos: RwLockWriteGuard` from earlier in this fn, and I don't like the idea of passing around lock guards to functions.

Comment (andriyDev) [reply crates/bevy_asset/src/server/mod.rs:422]: Can you just take a mutable borrow to the thing it's guarding, which I believe should be `AssetInfos`? Then you can pass that around with the guard.

Comment (aecsocket) [reply crates/bevy_asset/src/server/mod.rs:422]: Yes, but I also don't like the idea of locking, working, unlocking, then relocking again for some other work in the same function. Is this an ok tradeoff for less code duplication?

Comment (andriyDev) [reply crates/bevy_asset/src/server/mod.rs:422]: I don't think you need to, you can just "unwrap" the `infos` like `&mut *infos`, and that'll give you `&mut AssetInfos` while still being protected by the guard. Then you can pass that borrow to the function.

Comment (aecsocket) [reply crates/bevy_asset/src/server/mod.rs]: Yep you're right, I forgot you could just pass the mut ref.

**Review by alice-i-cecile — APPROVED** [review]
Fun use of the type state pattern (and `Either`), clear motivation, and great docs. I like this!

**Comment by JMS55** [comment]
Sorry, realistically I'm not going to have time to review this.

File: crates/bevy_asset/src/server/info.rs [hunk 7]
```diff
@@ -162,15 +163,15 @@ impl AssetInfos {
         );
         // it is ok to unwrap because TypeId was specified above
         let (handle, should_load) =
-            unwrap_with_context(result, core::any::type_name::<A>()).unwrap();
+            unwrap_with_context(result, Either::Left(core::any::type_name::<A>())).unwrap();
```
Comment (bushrat011899) [inline crates/bevy_asset/src/server/info.rs:166]: I'm _personally_ not a fan of `Either`. I'd rather just define a little `enum` with descriptive names (rather than just `Left` and `Right`), and take the opportunity to implement `From<...>` for each of the possibilities, since we _know_ what each type option is (and thus won't need specialisation).

```rust
pub(crate) enum MyTypeInfo {
    Name(&'static str),
    TypeId(TypeId),
}

impl From<&'static str> for MyTypeInfo { ... }
impl From<TypeId> for MyTypeInfo { ... }

pub(crate) fn unwrap_with_context<T>(
    result: Result<T, GetOrCreateHandleInternalError>,
    type_info: impl Into<MyTypeInfo>,
) -> Option<T> { ... }

unwrap_with_context(result, core::any::type_name::<A>());
unwrap_with_context(result, core::any::TypeId::of::<A>());
```

Having said that, I know not everyone agrees with me on this opinion and I _do_ think this is one of the best examples of where to appropriate use `Either`. I won't hold-up merging just for my own personal preferences!

**Review by bushrat011899 — APPROVED** [review]
I think this is a good API change and very well documented. I do have a couple of suggestions but nothing I'd consider blocking.

**Comment by alice-i-cecile** [comment]
@aecsocket I'll give you a day or two to respond to the review, then merge this in before the release candidate ships :)

File: crates/bevy_asset/src/loader_builders.rs [hunk 8]
```diff
@@ -74,47 +209,105 @@ impl<'ctx, 'builder> NestedLoader<'ctx, 'builder> {
...
+            load_context: self.load_context,
+            meta_transform: self.meta_transform,
+            typing: UnknownTyped { _priv: () },
+            mode: self.mode,
+        }
+    }
+
+    // convert between `M`s
+
+    /// When [`load`]ing, create only asset handles, rather than loading the
```
Comment (andriyDev) [inline crates/bevy_asset/src/loader_builders.rs:266]: Perhaps "rather than *returning* the actual asset". It sounds a little strange, like we don't load the asset at all with deferred.

File: crates/bevy_asset/src/loader_builders.rs [hunk 9]
```diff
@@ -25,28 +25,163 @@ impl ReaderRef<'_> {
...
             load_context,
             meta_transform: None,
-            asset_type_id: None,
+            typing: StaticTyped { _priv: () },
+            mode: Deferred { _priv: () },
         }
     }
+}
 
+impl<'ctx, 'builder, T: sealed::Typing, M: sealed::Mode> NestedLoader<'ctx, 'builder, T, M> {
```
Comment (andriyDev) [inline crates/bevy_asset/src/loader_builders.rs:184]: One thing that's a little odd to me, I'm not sure if we care enough: with this impl, you can say:

```rust
load_context.loader().with_static_type().with_static_type().with_static_type()...
```

We could solve this with separate impl blocks, but perhaps that's too much boilerplate.

Feel free to ack!

File: crates/bevy_asset/src/server/info.rs [hunk 10]
```diff
@@ -181,7 +182,11 @@ impl AssetInfos {
             meta_transform,
         );
         // it is ok to unwrap because TypeId was specified above
```
Comment (andriyDev) [inline crates/bevy_asset/src/server/info.rs:184]: This comment is in the wrong spot. Alternatively, maybe just making this an expect is appropriate?

**Review by andriyDev — APPROVED** [review]
A couple minor changes, nothing critical!

File: crates/bevy_asset/src/loader_builders.rs (hunk 9 above)
Comment (BenjaminBrienen) [reply crates/bevy_asset/src/loader_builders.rs:184]: I think this small improvement would be nice. I bet you someone will notice it and report an issue anyway!

File: crates/bevy_asset/src/loader_builders.rs [hunk 11]
```diff
@@ -25,28 +25,163 @@ impl ReaderRef<'_> {
...
     meta_transform: Option<MetaTransform>,
-    asset_type_id: Option<TypeId>,
+    typing: T,
+    mode: M,
+}
+
+mod sealed {
+    pub trait Typing {}
+
+    pub trait Mode {}
```
[the remark this answers is not in the document]
Comment (aecsocket) [reply crates/bevy_asset/src/loader_builders.rs:128]: Agreed, the two "modes" you outlined are the only ones we really should support. There's no real point in having the other 2.

File: crates/bevy_asset/src/server/info.rs (hunk 7 above)
Comment (aecsocket) [reply crates/bevy_asset/src/server/info.rs:166]: I agree that this could be more ergonomic, using something like
```rs
enum TypeNameOrId<'a> {
    Name(&'a str),
    Id(TypeId),
}
fn foo(type_info: impl Into<TypeNameOrId<'_>>) { .. }
```
However, I'd like to leave that for a future PR, since I bet I could find some more places where this would be a nice ergonomic change, which isn't exclusive to this PR's NestedLoader API changes.
Relevant discussion here: https://discord.com/channels/691052431525675048/749332104487108618/1290249951556665444
I'll keep it as Either for now

File: crates/bevy_asset/src/loader_builders.rs [hunk 12]
```diff
@@ -25,28 +25,163 @@ impl ReaderRef<'_> {
...
+}
+
+/// [`NestedLoader`] will be provided the type of asset as a type parameter on
+/// [`load`].
+///
+/// [`load`]: NestedLoader::load
+pub struct StaticTyped {
+    _priv: (),
```
[the remark this answers is not in the document]
Comment (aecsocket) [reply crates/bevy_asset/src/loader_builders.rs:136]: No real advantage, I like using a struct with named fields this kind of private-only unit struct since the `_priv` makes it super obvious why we need that field.

However, it does mean that initialization has to be spread over multiple lines, which is kind of ugly, so I think I'll change it to a private unit field in a tuple struct.

File: crates/bevy_asset/src/loader_builders.rs (hunk 9 above)
Comment (aecsocket) [reply crates/bevy_asset/src/loader_builders.rs:184]: To enable this for at least just the `T` parameter, I'd need:
```rs
impl<M> NestedLoader<StaticTyped, M> {
    fn with_dynamic_type() -> .. {}
    fn with_unknown_type() -> .. {}
}

impl<M> NestedLoader<DynamicTyped, M> {
    fn with_static_type() -> .. {}
    fn with_unknown_type() -> .. {}
}

impl<M> NestedLoader<UnknownType, M> {
    fn with_static_type() -> .. {}
    fn with_dynamic_type -> .. {}
}
```
I really don't like the duplication here, but what's a deal breaker for me with this approach IMO is I would have to duplicate docs for both of the `with_x_type` etc. Docs are so much harder to keep up to date than code, and I'd always prioritise having a single source of truth for docs over duplicating docs. Inevitably, if we do this, someone will change how these functions work and change the docs for one of these, then forget the other.

I could also add some sealed traits like `TypedNotStatic`, `TypedNotDynamic`, `TypedNotUnknown`, etc. and impl for those, but IMO this is over-engineering a solution to something that isn't really a problem. I think if someone writes
```rs
my_loader
    .with_static_type()
    .with_static_type()
```
it's not really harmful/a footgun of any kind, and it's easily spotted in code review.

The same problem exists for `M` as well, but to a lesser extent.

File: crates/bevy_asset/src/server/info.rs (hunk 10 above)
Comment (aecsocket) [reply crates/bevy_asset/src/server/info.rs]: Changed it to an expect, since that's why the stdlib even has that function in the first place

**Comment by aecsocket** [comment]
Resolved everything apart from 1 comment. I think my reasoning makes sense for leaving that one open.
